src/dump.c

```diff
@@ -290,7 +294,38 @@ static void jl_serialize_datatype(jl_serializer_state *s, jl_datatype_t *dt) JL_
 {
     int tag = 0;
     int internal = module_in_worklist(dt->name->module);
-    if (!internal && jl_unwrap_unionall(dt->name->wrapper) == (jl_value_t*)dt) {
+    if (mini_image && !jl_is_tuple_type(dt) && !jl_is_array_type(dt) && dt != jl_float64_type) {
+        if (dt->layout && jl_datatype_nfields(dt) > 0) {  // change the type to a tuple type with correct size
+            dt->name->module = jl_core_module;  // fudge the module
+            // jl_datatype_t *newdt = jl_apply_tuple_type(dt->types);
+            // newdt->layout = dt->layout;
+            // newdt->instance = dt->instance;
+            // newdt->ninitialized = dt->ninitialized;
+            // dt = newdt;
+        }
+        else if (dt->size == 1) {  // change the type to a UInt8
+            dt = jl_uint8_type;
+        }
+        else if (dt->size == 2) {  // change the type to a UInt16
+            dt = jl_uint16_type;
+        }
+        else if (dt->size == 4) {  // change the type to a UInt32
+            dt = jl_uint32_type;
+        }
+        else if (dt->size == 8) {  // change the type to a UInt64
+            dt = jl_uint64_type;
+        }
+        else if (dt->size > 0) {  // change the type to a primitive type with correct size
+            dt = jl_new_primitivetype(jl_symbol("BitsTypeX"), jl_core_module, jl_any_type, jl_emptysvec, dt->size * 8);
```

Comment:
If this code doesn't have a fixed repertoire of types --- such that it can handle this new primitive type --- then why not just save the correct type to begin with?